Sources/Smithy/Logging/LogAgent.swift

```diff
@@ -23,9 +23,9 @@ public protocol LogAgent {
     ///     - function: The function the log line was emitted from.
     ///     - line: The line the log message was emitted from.
     func log(level: LogAgentLevel,
-             message: String,
-             metadata: [String: String]?,
-             source: String,
+             message: @autoclosure () -> String,
+             metadata: @autoclosure () -> [String: String]?,
+             source: @autoclosure () -> String,
```

Comment:
`message`, `metadata`, and `source` are now marked `@autoclosure`.

https://docs.swift.org/swift-book/documentation/the-swift-programming-language/closures/#Autoclosures

This prevents the parameter from being evaluated unless it is actually used by the function being called.  In the context of logging, this means that these params are not constructed when a log message is not going to be generated (i.e. when the log level of a statement is below the logger's threshold.)

This corresponds with `swift-log`'s `LogHandler` protocol, where these three params are also autoclosures.

(Note that `#file`, `#function`, and `#line` evaluate to static strings, so there is no real benefit to making them autoclosures.)